Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,194 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from fontTools import ttLib
+from fontTools.pens.hashPointPen import HashPointPen
+from fontTools.ttLib.tables._g_l_y_f import (
+    OVERLAP_COMPOUND,
+    ROUND_XY_TO_GRID,
+    USE_MY_METRICS,
+)
+from ufo2ft.constants import (
+    TRUETYPE_INSTRUCTIONS_KEY,
+    TRUETYPE_ROUND_KEY,
+    TRUETYPE_METRICS_KEY,
+    TRUETYPE_OVERLAP_KEY,
+    OBJECT_LIBS_KEY,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class InstructionCompiler(object):
+    def __init__(self, ufo, ttf):
+        self.ufo = ufo
+        self.font = ttf
+
+    def _compile_program(self, key, table_tag):
+        assert table_tag in ("prep", "fpgm")
+        ttdata = self.ufo.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata:
+            formatVersion = ttdata.get("formatVersion", None)
+            if int(formatVersion) != 1:
+                logger.error(
+                    f"Unknown formatVersion {formatVersion} "
+                    f"in key '{key}', "
+                    f"table '{table_tag}' will be empty in font."
+                )
+                return
+            asm = ttdata.get(key, None)
+            if asm is not None:
+                self.font[table_tag] = table = ttLib.newTable(table_tag)
+                table.program = ttLib.tables.ttProgram.Program()
+                table.program.fromAssembly(asm)
+                # Roundtrip once, or if the font is dumped to XML before having
+                # been saved, the assembly code if will look awful.
```

Comment:
```suggestion
                # been saved, the assembly code will look awful.
```

how "awful"? this sounds like a bug in fonttools?

Reply:
IIRC each PUSH instruction's arguments will be on one line each instead of all on the same line.

Reply:
I think I noticed that when saving a ttx file at this point while debugging. For normal operation it shouldn't matter. I'll remove it.